Source/AssetRipper.Processing/AnimatorControllers/AnimatorStateContext.cs

```diff
@@ -0,0 +1,190 @@
+using AssetRipper.Assets.Collections;
+using AssetRipper.Assets.Generics;
+using AssetRipper.SourceGenerated.Classes.ClassID_1102;
+using AssetRipper.SourceGenerated.Classes.ClassID_91;
+using AssetRipper.SourceGenerated.Extensions;
+using AssetRipper.SourceGenerated.Subclasses.StateConstant;
+using AssetRipper.SourceGenerated.Subclasses.StateMachineConstant;
+
+namespace AssetRipper.Processing.AnimatorControllers;
+
+public sealed class AnimatorStateContext
+{
+	public readonly int StateCount;
+	public bool HasStates() => StateCount > 0;
+	public readonly int DefaultStateIdx;
+
+	private readonly ProcessedAssetCollection VirtualFile;
+	private readonly IAnimatorController Controller;
+	private readonly IStateMachineConstant StateMachineConstant;
+	private readonly int LayerIndex;
+	
+	private readonly IAnimatorState[] states;
+	private readonly IStateConstant[] stateConstants;
+	private readonly uint[] stateIdxsToStateMachinePathIDs; // for grouping States into StateMachines
+	private readonly BidirectionalDictionary<string,uint> stateMachinePathNamesAndIDs;
+
+	public AnimatorStateContext(ProcessedAssetCollection virtualFile, IAnimatorController controller, IStateMachineConstant stateMachineConstant, int layerIndex)
+	{
+		VirtualFile = virtualFile;
+		Controller = controller;
+		StateMachineConstant = stateMachineConstant;
+		LayerIndex = layerIndex;
+
+		DefaultStateIdx = stateMachineConstant.DefaultState != uint.MaxValue ? (int)stateMachineConstant.DefaultState : 0;
+		StateCount = stateMachineConstant.StateConstantArray.Count;
+
+		stateMachinePathNamesAndIDs = new();
+		if (!HasStates())
+		{
+			stateConstants = [];
+			states = [];
+			stateIdxsToStateMachinePathIDs = [];
+			return;
+		}
+		stateConstants = new IStateConstant[StateCount];
+		states = new IAnimatorState[StateCount];
+		stateIdxsToStateMachinePathIDs = new uint[StateCount];
+	}
+
+	/// <summary>
+	/// Create all AnimatorStates, and group them by StateMachine paths
+	/// </summary>
+	public void Process()
+	{
+		if (!HasStates()) { return; }
+
+		if (!Controller.TOS.ContainsKey(0))
+		{
+			Controller.TOS[0] = Utf8String.Empty;
+		}
```

Comment:
Can you move this functionality to `AssetDictionaryBase` as a `TryAdd` function?